File: stringLetra.py
def comprimir(cadena):
    if not cadena:
        return ""

    resultado = ""
    i = 0
    while i < len(cadena):
        caracter_actual = cadena[i]
        conteo = 1
        j = i + 1
        while j < len(cadena) and cadena[j] == caracter_actual:
            conteo += 1
            j += 1

        resultado += caracter_actual
        if conteo > 1:
            resultado += str(conteo)
        i = j
    return resultado

# Almacenar la palabra en una lista (opcional, el string ya es iterable)
palabra = "galleta"
lista_letras = list(palabra)
print(f"Lista de letras: {lista_letras}")

# Comprimir la cadena
cadena_comprimida = comprimir(palabra)
print(f"Cadena comprimida: {cadena_comprimida}")

# descomrpimir vector

import re

def descomprimir(cadena_comprimida):
    if not cadena_comprimida:
        return ""

    # Usa expresiones regulares para encontrar el patrón de letra seguida de números (opcional)
    # Ejemplo: 'a2' o 'b'
    matches = re.findall(r'([a-zA-Z])(\d*)', cadena_comprimida)
    
    resultado = ""
    for caracter, conteo_str in matches:
        conteo = int(conteo_str) if conteo_str else 1
        resultado += caracter * conteo
        
    return resultado

cadena_descomprimida = descomprimir(cadena_comprimida)
print(f"Cadena descomprimida: {cadena_descomprimida}")

File: stringLetra.js
//almacenar cada letra de la palabra "galleta" en un arreglo

console.log("---- el string escogido es galleta  ----");

const palabra = "galleta"
const arregloLetras = palabra.split("");

console.log(arregloLetras);

//comprimir cadena de letras

console.log("---- vector de letras repetidas comprimido ----");

function comprimir(cadena) {
  if (!cadena) {
    return "";
  }
  let resultado = "";
  let contador = 1;

  for (let i = 0; i < cadena.length; i++) {
    if (i + 1 < cadena.length && cadena[i] === cadena[i + 1]) {
      contador++;
    } else {
      resultado += cadena[i] + (contador > 1 ? contador : "");
      contador = 1;
    }
  }
  return resultado;
}

const cadenaOriginal = arregloLetras;
const cadenaComprimida = comprimir(cadenaOriginal);

console.log(cadenaComprimida);

// descomprimir string galleta

console.log("---- string descomprimido ----");

function descomprimir(cadenaComprimida) {
  if (!cadenaComprimida) {
    return "";
  }
  let resultado = "";
  let i = 0;

  while (i < cadenaComprimida.length) {
    let caracter = cadenaComprimida[i];
    i++;
    let numeroStr = "";
    while (i < cadenaComprimida.length && !isNaN(parseInt(cadenaComprimida[i]))) {
      numeroStr += cadenaComprimida[i];
      i++;
    }
    const repeticiones = numeroStr ? parseInt(numeroStr) : 1;
    resultado += caracter.repeat(repeticiones);
  }
  return resultado;
}

const cadenaComprimida2 = cadenaComprimida;
const cadenaDescomprimida = descomprimir(cadenaComprimida2);

console.log(cadenaDescomprimida);


